keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -0,0 +1,578 @@
+nodes are added or removed, the proportion of endpoints that should be allocated
+to each zone will change. This will be especially common in autoscaling
+scenarios.
+
+To mitigate the number of changes resulting from these events, EndpointSlices
+will only be updated if a Node addition or removal results in a transition above
+or below the overload threshold. For example, syncs would be triggered in either
+of the following scenarios:
+
+1. A deleted Node results in a Service exceeding the overload threshold.
+2. A new Node results in a Service that is able to achieve an endpoint
+   distribution below 20% for the first time.
+
+### Future Expansion
+
+In the future we may expand this functionality if needed. This could include:
+
+- A new `RequireZone` algorithm that would keep endpoints in EndpointSlices for
+  the same zone they are in.
+- A new option to specify a minimum threshold for the `Auto` approach.
+- Support for region based hints.
+
+### Test Plan
+
+#### Controller Unit Tests
+| Test Description | Expected Result |
+| :--- | :--- |
+| Feature Gate On, Annotation == 'Auto', 2+ zones | Hints set |
+| Feature Gate On, Annotation == 'Auto', 1 zone | No hints set |
+| Feature Gate On, Annotation == 'On', 2+ zones | No hints |
+| Feature Gate On, Annotation Off, 2+ zones | No hints |
+| Feature Gate Off, Annotation On, 2+ zones | No hints |
+| Feature Gate Off, Annotation Off, 2+ zones | No hints |
+| Feature Gate Off, Annotation Off, 2+ zones | No hints |
+| 2 endpoints, 3 zones | No hints |
+| 3 endpoints, 3 zones | Hints set |
+| 4 endpoints, 3 zones | No hints |
+| 4 endpoints, 2 zones | Hints set |
+| 4 endpoints all from 1 zone, 2 zones | Hints set |
+| 4 endpoints, 3 zones, 1 zone with 2x cores | Hints set |
+| 400 endpoints, 4 zones with slightly different cores | Hints set |
+| Node removal that does not trigger threshold transition | No EndpointSlice changes |
+| Node removal that triggers threshold transition | EndpointSlice updates |
+| Node without way to determine cores | All Nodes treated equally |
+| Endpoint additions that require redistribution | Hints updated |
+| Endpoint removals that require redistribution | Hints updated |
+
+#### Kube-Proxy Unit Tests
+| Test Description | Expected Result |
+| :--- | :--- |
+| Feature Gate On, Annotation == 'Auto', hints matching zone | Endpoints filtered |
+| Feature Gate On, Annotation == 'On', hints matching zone | Endpoints not filtered |
+| Feature Gate Off, Annotation == 'Auto', hints matching zone | Endpoints not filtered |
+| Feature Gate On, Annotation == 'Auto', no hints matching zone | Endpoints not filtered |
+
+### Observability
+We can reuse some of the metrics of EndpointSlice Controller that we already
+have in the current version to observe the changes of endpoints (addition,
+deletion and update). Meanwhile we can add more metrics to have a glimpse of
+different approaches.
+
+- `endpoint_slice_controller/endpointslices_changed_per_sync`
```

Comment:
it seems to me that the overload threshold and metrics per zone it is going to be something important to have insights.

Reply:
Agree that this would be useful information to have. I'm not quite sure how to represent it yet, but will work on something.

Reply:
Added a new `endpoint_overload_by_zone` metric - thanks for the idea!